olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
+
+OperatorBundle
+    -name: foo
+    -bundle: NULL
+    -bundlePath: "quay.io/community-operators/foo:sha256@abcd123"
+    
+Additionally, we will update the grpc API that OLM currently uses to get the bundle that if it attempts to pull the non latest version will return null. In that case, it will make a second query to get the bundlePath. Then we will create a pod with that image that will serve the bundle data by writing the data to a configmap that can be read by OLM.
+
+In order to serve that manifest image, we cannot pull the image directly. Instead we will:
+
+1. Create a pod spec that uses an init container to inject a binary into the manifest image.
```

Comment:
I may not be understanding this correctly.  Is this a proposal to run a pod for each operator we are gathering metadata for?

Reply:
The premise is that we need some way of getting the manifests stored in these bundle images. We are going to accomplish it by bootstrapping a pod to push the metadata from it into a configmap and then kill the pod - we aren't going to run these pods perpetually.

This is a (non ideal) method of using normal kubernetes apis to pull metadata files from an image that is just scratch + files copied into a directory. In an ideal world we could use container tooling in olm directly to pull the image and extract these files from the image, but that is not an option with what we have available to us now. @jwforres @shawn-hurley